### PR TITLE
Fix gpt2 sample

### LINKED ISSUE #26517: GPT2 sample failing with the new graph engine
### System Information

OpenCV version: 5.x
Operating System / Platform: MacOS

### Detailed description

Running gpt2 inference sample fail if the sequence length is different comapred to export  sequence length. This behaviour is present using new graph engine (it should suppport dynamic shapes). 

Error is 
```bash
logits = net.forward()
cv2.error: OpenCV(5.0.0-pre) /Users/abd/work/opencv/modules/dnn/src/layers/nary_eltwise_layers.cpp:308: error: (-215:Assertion failed) shape[i] == 1 || outShape[i] == 1 in function 'findCommonShape'
```

### Steps to reproduce

Run[ gpt2 inference sample](https://github.com/opencv/opencv/blob/5.x/samples/dnn/gpt2_inference.py) as described in the sample. Include [this patch](https://github.com/opencv/opencv/pull/26514#issuecomment-2497024212) for the running the sample with new engine

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### LINKED ISSUE #26518: Errors for gpt2_inference after 5.x dnn update
### System Information

My system info:

```
env: WSL2-Ubuntu20.04
```

Related package:

```
Python 3.8.10
pip 24.3.1
numpy 1.24.3
tiktoken 0.6.0
opencv 5.0.0
```

I build up `opencv 5.x` on these steps:

```bash
git clone --branch 5.x https://github.com/opencv/opencv.git
cmake -B opencv-build -S opencv 
    -DCMAKE_INSTALL_PREFIX=opencv-build/install 
    -DWITH_TIFF=OFF 
    -DWITH_IPP=OFF 
    -DWITH_JPEG=OFF 
    -DWITH_AVIF=OFF 
    -DCMAKE_BUILD_TYPE=Release
cmake --build opencv-build --target install
python3 -m pip install opencv-build/python_loader
```

### Detailed description

Before the dnn module update, the sample of gpt2_inference ran normally. But after update, some errors occur. Here's some of the problems: 

1. `net.getLayerShapes` **changes its return type**. The source code can be seen in `5.x/samples/dnn/gpt2_inference.py: Line105`

```python
net = cv.dnn.readNet(args.model)
input_token_size = net.getLayerShapes([], 0, 0)[0][0][1]  # Problem's here!
```

- When trying to get input size, the last dimension returns a `cv2.MatShape` value. And I was told that `'cv2.MatShape' object is not subscriptable` .

2. `net.setInput` **expects** `CV_64SC1`, but Python would convert the input `x` to `CV_32SC1` and raise an error. The source code can be seen in `5.x/samples/dnn/gpt2_inference.py: Line67`

```python
def gpt2_inference(net, tokens, max_length, num_return_sequences=5):
    # ...
    x = np.array(tokens)
    x = np.tile(x, (num_return_sequences, 1))
    while counter < max_length:
        net.setInput(x)  # Problem's here!
```

- And the program was interrupted here, so I have not found other problems yet.
- Is there any problems here? Or just because the opencv-python API has not been updated completely?

### Steps to reproduce

`samples/dnn/gpt2_inference.py` has the same problem. 

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by Abdurrahheem**
This PR adds dynamic input support for `gpt2_inference.py` sample.
Fixes #26518
Fixes #26517

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by asmorkalov — APPROVED** [review]

**Comment by asmorkalov** [comment]
tested manually. Works well! Thanks for the great job!
